### PR TITLE
Product Query - Enable "Inherit Query from template" option

### LINKED ISSUE #7629: Product Query [3]: [Tests] Add E2E tests for the FSE functionalities
E2E tests should test **at least** these paths:

### Editor side:

* When the toggle is enabled, all filters are disabled and can't be chosen anymore. If they were chosen before, they are now reset.
* Toggle should not be available inside non-FSE pages and posts.

### Front-end side:

* It works in combination with the Product Search template.
* It works in combination with the Products by Category template.
* It works in combination with the Product Catalog template.
* It works in combination with the Products by Tag template.

Note that this issue is not supposed to E2E test filters.

## Old specs

These specs only apply if we do a custom implementation of “Inherit Query from Template”:

### Editor side:

* When the toggle is enabled, taxonomies are disabled and can't be chosen anymore from the filters. If they were chosen before, they are now reset.
* When the toggle is enabled, and we are in FSE templates, the correct preview is shown.

### Front-end side:

* More specific settings can override the global query.

### PR DESCRIPTION
**Pull request by gigitux**
This PR restores the "Inherit Query from template" from Gutenberg. The setting option will be visible only when the block is added in the Site Editor. Furthermore, at least for now, when the setting is enabled, the rest of the settings that customize the query are hidden. 

Also, this PR could be considered the starting point for working on #7629. These E2E tests check that the results between the Classic Templates and the Product Query are equal (same products, in the same order).

Fixes #7629

> **Warning**
> The UI is not definitive.

## Post/Page Editor 
image

## Site Editor
https://user-images.githubusercontent.com/4463174/206170785-77b9b1d8-2d0c-45ea-a20e-76ddae1b4f74.mp4

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

## Site Editor
1. Add the Product Query in a template (eg: Product Catalog)
2. Check that the `Inherit Query from template` option is visible.
3. Enable it.
4. Be sure that all the settings that customize the query are hidden.
5. Save.
6. Be sure that the result generated by the Product Query is the same as the classic FSE template (there are E2E tests that check this ).
7. Edit the Product Catalog and disable the `Inherit Query from template` option.
8. Ensure all the settings that customize the query are visible.

## Post/Page Editor

1. Add the Product Query in a template (eg: Product Catalog)
2. Check that the `Inherit Query from template` option isn't visible.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

> Product Query: Add the "Inherit Query from template" option.

### CODE REVIEW

File: tests/e2e/config/jest.config.js
Comment (gigitux) [inline tests/e2e/config/jest.config.js]: The first two lines are the default from the jest configuration (source: https://jestjs.io/docs/28.x/configuration#testmatch-arraystring).
The last one is for not running jest on all the files that have the suffix `-utils` or called `utils`.

File: tests/e2e/specs/shopper/product-query/utils.ts
Comment (Aljullu) [inline tests/e2e/specs/shopper/product-query/utils.ts]: Are these two selectors used somewhere?

File: tests/e2e/config/jest.config.js
Comment (Aljullu) [reply tests/e2e/config/jest.config.js]: Could the last line be included in [`testPathIgnorePatterns`](https://jestjs.io/docs/configuration#testpathignorepatterns-arraystring) so we don't need `testMatch`?

File: tests/e2e/specs/shopper/product-query/product-query-with-templates.test.ts
Comment (Aljullu) [inline tests/e2e/specs/shopper/product-query/product-query-with-templates.test.ts]: I would prefer if we compared the arrays directly:

```suggestion
			expect( classicProducts ).toEqual( products );
```

For this to work, however, we would need to move the `afterAll()` from L25 to inside each `describe()`. And then, that causes the error @albarin fixed in https://github.com/WordPress/gutenberg/pull/45992. Related issue: https://github.com/woocommerce/woocommerce-blocks/issues/7744.

So I think we can move ahead without making the change, but it would be good to note that in #7744 so we don't forget. How does it sound?

**Review by Aljullu — APPROVED** [review]
I left a couple of comments/questions, but overall looking good, so pre-approving.

**Comment by gigitux** [comment]
Thanks for the first review, @Aljullu.
After the last discussion, I increased the scope of this PR, and I implemented a custom `inherit query from the template` option.
Could you take another look? 🙇

**Review by Aljullu — APPROVED** [review]
LGTM, thanks for working on this complex PR, @gigitux!

One question: should _WooCommerce Visibility_ in the PR description be updated to _WooCommerce Core_?

Also, not a blocker but I found an unexpected situation:

* Edit the _Product Search Results_ template.
* Add the _Products_ block.
* Add the _Sale Status_ filter and select _Show only products on sale_.
* Now, enable _Inherit query from template_.
* Notice only products on sale are displayed in the editor.
* Instead, all products are shown in the frontend.

As said, I don't think it's a blocker because in the frontend it works fine, but it might be good to create a follow-up issue.

**Comment by gigitux** [comment]
@vivialice @nerrad Should we improve the labels before merging this PR?

image

**Comment by nerrad** [comment]
> @vivialice @nerrad Should we improve the labels before merging this PR?

How trivial is it to change the labels or description?

**Comment by gigitux** [comment]
> > @vivialice @nerrad Should we improve the labels before merging this PR?
> 
> How trivial is it to change the labels or description?

Super easy! It is necessary updates the strings [here](https://github.com/woocommerce/woocommerce-blocks/pull/7641/files#diff-7da64db1c9a7cb730987449b179e69361060de41b0bc1a9a9cfbc4bb28fb40fbR157-R163).

**Comment by nerrad** [comment]
Hmm, I'm on the fence about changing the wording. As is, it concisely explains technically what the option does. I'd prefer to roll out with minimal changes for now until we see how much of an issue it is.

So I would suggest changing just the description for now:
| Before | After |
| ---- | ---- |
| Toggle to use the global query context that is set with the current template, such as an archive or search. Disable to customize the settings independently. | Toggle to use the global query context that is set with the current template, such as variations of the product catalog or search. Disable to customize the filtering independently. |

**Comment by gigitux** [comment]
> LGTM, thanks for working on this complex PR, @gigitux!
> 
> One question: should _WooCommerce Visibility_ in the PR description be updated to _WooCommerce Core_?
> 
> Also, not a blocker but I found an unexpected situation:
> 
> * Edit the _Product Search Results_ template.
> * Add the _Products_ block.
> * Add the _Sale Status_ filter and select _Show only products on sale_.
> * Now, enable _Inherit query from template_.
> * Notice only products on sale are displayed in the editor.
> * Instead, all products are shown in the frontend.
> 
> As said, I don't think it's a blocker because in the frontend it works fine, but it might be good to create a follow-up issue.

Nice catch! The preview for other templates is updated correctly because [Gutenberg logic takes care of it](https://github.com/WordPress/gutenberg/blob/7dda7859adfa06adbe0da026f8f1253c2b001847/packages/block-library/src/post-template/edit.js/#L168-L177).

This doesn't happen with the Product Search Results because its slug doesn't start with `archive-` or `category-`.  
We can easily fix this by restoring the default query.
The only disadvantage of this approach is that if the user applies some filters, enables the inherit query from the template option, and after that disables it, will lose all the customization that did.

**Comment by gigitux** [comment]
> Hmm, I'm on the fence about changing the wording. As is, it concisely explains technically what the option does. I'd prefer to roll out with minimal changes for now until we see how much of an issue it is.
> 
> So I would suggest changing just the description for now:
> 
> Before	After
> Toggle to use the global query context that is set with the current template, such as an archive or search. Disable to customize the settings independently.	Toggle to use the global query context that is set with the current template, such as variations of the product catalog or search. Disable to customize the filtering independently.

Update the label with https://github.com/woocommerce/woocommerce-blocks/pull/7641/commits/49be1e79990b84de2baa308e855da62597b52bbb.

**Comment by Aljullu** [comment]
>  We can easily fix this by restoring the default query.
>  The only disadvantage of this approach is that if the user applies some filters, enables the inherit query from the template option, and after that disables it, will lose all the customization that did.

Yeah, I prefer the current behavior than the user losing the attributes after toggling on and off the inherit query from the template option.

For this specific PR, I would merge it as-is. In the future, we can investigate if there are alternative approaches to reset the query in the editor, but definitely not a blocker.

**Comment by gigitux** [comment]
I merged the PR and I created a follow-up issue #7921.

**Comment by vivialice** [comment]
> Hmm, I'm on the fence about changing the wording. As is, it concisely explains technically what the option does. I'd prefer to roll out with minimal changes for now until we see how much of an issue it is.

Do you think this may be too technical @nerrad ? I am okay with your suggestion but I feel like the average user might not understand what a `global query context` is. Perhaps we can think about including some messaging around it being a "recommended" setting?

**Comment by sunyatasattva** [comment]
+1 On the copy being a tad too technical and esoteric, both of the label and the help text. Summoning @ifinoi 😁

**Comment by nerrad** [comment]
>  I am okay with your suggestion but I feel like the average user might not understand what a global query context is.

I don't disagree, I just haven't thought of more friendly wording to describe what this toggle does. 

> Perhaps we can think about including some messaging around it being a "recommended" setting?

Maybe? But then why is it recommended? 

If we can come up with some more user-friendly wording that still correctly describes what this toggle does, I'm all for changing it, just not sure it needs to block the release to core while still beta. I also want to avoid too much churn around translated strings until we've settled on something.

**Comment by nerrad** [comment]
Something else that I waffle between is that while we _are_ aiming to make the store editing experience in the context of the Site editor as merchant friendly as possible, I think there are still areas where more technical terms are the best way of describing what certain functionality does. Things like "query", "template", "filter" are words used to describe behaviour that often doesn't have really great similes and descriptions can get fairly wordy fast (which pollutes the UI). At some point, the concept of "this template uses data from the context of the route" needs to be understood on some level when modifying the template's layout.

What we _could_ do, is for those concepts that are more technical than others, we could link off to documentation that uses more words to describe what a query is, or further expands on a description.

**Comment by nerrad** [comment]
Some good reference material around the topic of phrasing and user experience in general for the WP Query Loop block are the following issues in the Gutenberg project:

- https://github.com/WordPress/gutenberg/issues/40944
- https://github.com/WordPress/gutenberg/issues/41405

**Pull request merged**
